src/topology/metric_space/lipschitz.lean

```diff
@@ -44,6 +54,44 @@ by { simp only [lipschitz_with, edist_nndist, dist_nndist], norm_cast }
 
 alias lipschitz_with_iff_dist_le_mul ↔ lipschitz_with.dist_le_mul lipschitz_with.of_dist_le_mul
 
+/-- A function `f` is Lipschitz continuous with constant `K ≥ 0` on `s` if for all `x, y` in `s`
+we have `dist (f x) (f y) ≤ K * dist x y` -/
+def lipschitz_on_with [emetric_space α] [emetric_space β] (K : ℝ≥0) (s : set α) (f : α → β) :=
+∀ x y ∈ s, edist (f x) (f y) ≤ K * edist x y
```

Comment:
Note that `∀ (x ∈ s) (y ∈ s)` works better if you have a lemma simplifying `∀ x ∈ s, p x`.

Reply:
I'd add `lipschitz_on_with K s f ↔ lipschitz_with K (s.restrict f)`. Then, e.g., `lipschitz_on_with.uniform_continuous_on` should follow from `lipschitz_with.uniform_continuous`.

Reply:
I just tried to switch to `∀ (x ∈ s) (y ∈ s)` but I can't say I'm convinced, especially since I don't know how to make `x` and `y` implicit using this syntax (leaving only the proofs of membership explicit). But more seriously, it made the proof of `lipschitz_on_with K s f ↔ lipschitz_with K (s.restrict f)` into a nightmare. I'd be very grateful if you can golf it naturally.

Reply:
Unfortunately, making `x` and `y` implicit is harder with this syntax: you need something like `∀ {{x}} (hx : x ∈ s) {{y}} (hy : y ∈ s), p x y`. I hope, we'll be able to add a good syntax for this in Lean 4. As for the proof, I golfed it using recently added `set_coe.forall'`.

Reply:
Thanks, I missed `set_coe.forall'`. It's a bit weird that `set_coe.forall` is the simp lemma and `set_coe.forall'` is the useful one. @sgouezel could you merge that now that Yury is also an author?

Reply:
I added `set_coe.forall'` in #3760. It was merged after you wrote your proof.

Reply:
Oh I see. Nice timing!